File: structures/circular_linked_list/1.go
package main

import "fmt"

// 循环链表，设置头节点为哨兵节点

type Node struct {
	val  int
	next *Node
}
type NodeList struct {
	head *Node
}

func (n *NodeList) Insert(val int) {
	cur := n.head
	for cur.next != n.head && cur.next.val < val {
		cur = cur.next
	}
	newNode := &Node{
		val:  val,
		next: cur.next,
	}
	cur.next = newNode
}

func (n *NodeList) Search(val int) bool {
	cur := n.head
	for cur.next != n.head && cur.next.val < val {
		cur = cur.next
	}
	if cur.next != n.head && cur.next.val == val {
		return true
	} else {
		return false
	}
}

func (n *NodeList) Remove(val int) {
	cur := n.head
	for cur.next != n.head && cur.next.val < val {
		cur = cur.next
	}
	if cur.next != n.head && cur.next.val == val {
		cur.next = cur.next.next
	} else {
		fmt.Println("没有找到要删除的节点")
	}
}

func (n *NodeList) Print() {
	cur := n.head
	for cur.next != n.head {
		cur = cur.next
		fmt.Print(cur.val, "->")
	}
	fmt.Println("结束")
}

func main() {
	// 创建一个哨兵头节点，并让他指向自己
	sentinel := &Node{
		val:  -1,
		next: nil,
	}
	sentinel.next = sentinel
	list := &NodeList{
		head: sentinel,
	}
	list.Insert(1)
	list.Insert(2)
	list.Insert(5)
	list.Insert(3)
	list.Remove(1)

	a := list.Search(5)
	fmt.Println(a)

	list.Print()
}

File: structures/circular_linked_list/1.rs

#[derive(PartialEq,Clone)]
struct Node{
    val:i32,
    next:Option<Box<Node>>
}
struct NodeList{
    head:Option<Box<Node>>
}
// 循环链表，定义头节点为哨兵节点
impl NodeList{
    fn new()->NodeList{
        let mut sentinel=Box::new(Node{
            val:-1,
            next:None
        });
        sentinel.next=Some(sentinel.clone());
        NodeList{
            head:Some(sentinel)
        }
    }

    fn insert(&mut self, val: i32) {
        let mut cur = &mut self.head;
        // 遍历链表，找到合适的插入位置
        while let Some(node) = cur {
            if let Some(next_node) = &node.next {
                // 如果下一个节点的值小于要插入的值且不是哨兵节点，继续遍历
                if next_node.val < val && next_node.val != -1 {
                    cur = &mut node.next;
                    continue;
                }
            }
            // 找到合适的插入位置，创建新节点
            let new_node = Box::new(Node {
                val,
                next: node.next.take(),
            });
            // 将新节点插入到当前节点之后
            node.next = Some(new_node);
            return;
        }
    }
    fn print(&self){
        let mut cur=&self.head;
        while let Some(node)=cur{
            if node.next==self.head{
                break;
            }
            println!("{}",node.val);
            cur=&node.next;
        }
    }

}


fn main(){
    let mut list=NodeList::new();
    list.insert(1);
    list.insert(2);
    list.insert(3);
    list.print();

}